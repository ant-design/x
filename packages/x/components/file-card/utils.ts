export function getSize(size: number) {
  let retSize = size;
  const units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB'];
  let unitIndex = 0;

  while (retSize >= 1024 && unitIndex < units.length - 1) {
    retSize /= 1024;
    unitIndex++;
  }

  return `${retSize.toFixed(0)} ${units[unitIndex]}`;
}

export function matchExt(suffix: string, ext: string[]) {
  return ext.some((e) => suffix.toLowerCase() === `.${e}`);
}
