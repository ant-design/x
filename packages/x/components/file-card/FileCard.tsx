import {
  FileExcelFilled,
  FileImageFilled,
  FileMarkdownFilled,
  FilePdfFilled,
  FilePptFilled,
  FileTextFilled,
  FileWordFilled,
  FileZipFilled,
  JavaOutlined,
  JavaScriptOutlined,
  PythonOutlined,
} from '@ant-design/icons';
import { Image, type ImageProps } from 'antd';
import classnames from 'classnames';
import pickAttrs from 'rc-util/lib/pickAttrs';
import React, { useMemo } from 'react';
import useXComponentConfig from '../_util/hooks/use-x-component-config';
import { useXProviderContext } from '../x-provider';
import File from './components/File';
import AudioIcon from './icons/audio';
import VideoIcon from './icons/video';
import useStyle from './style';
import { matchExt } from './utils';

export type SemanticType = 'root' | 'file' | 'icon' | 'name' | 'description';
export type PresetIcons =
  | 'default'
  | 'excel'
  | 'image'
  | 'markdown'
  | 'pdf'
  | 'ppt'
  | 'word'
  | 'zip'
  | 'video'
  | 'audio'
  | 'java'
  | 'javascript'
  | 'python';

type FileExtendsProps =
  | ImageProps
  | Partial<React.JSX.IntrinsicElements['video']>
  | Partial<React.JSX.IntrinsicElements['audio']>;

export type FileCardProps = FileExtendsProps & {
  prefixCls?: string;
  style?: React.CSSProperties;
  styles?: Partial<Record<SemanticType, React.CSSProperties>>;
  className?: string;
  classNames?: Partial<Record<SemanticType, string>>;
  rootClassName?: string;
  key?: React.Key;
  name: string;
  byte?: number;
  size?: 'small' | 'default';
  description?: React.ReactNode;
  src?: string;
  mask?: React.ReactNode;
  icon?: React.ReactNode | PresetIcons;
  type?: 'file' | 'image' | 'audio' | 'video' | string;
  onClick?: () => void;
};

const IMAGE_EXT = ['png', 'jpg', 'jpeg', 'gif', 'bmp', 'webp', 'svg'];
const AUDIO_EXT = ['mp3', 'wav', 'flac', 'ape', 'aac', 'ogg'];
const VIDEO_EXT = ['mp4', 'avi', 'mov', 'wmv', 'flv', 'mkv'];

const PRESET_FILE_ICONS: {
  ext: string[];
  color: string;
  icon: React.ReactElement;
  key: string;
}[] = [
  {
    icon: <FileExcelFilled />,
    color: '#22b35e',
    ext: ['xlsx', 'xls'],
    key: 'excel',
  },
  {
    icon: <FileImageFilled />,
    color: '#8c8c8c',
    ext: IMAGE_EXT,
    key: 'image',
  },
  {
    icon: <FileMarkdownFilled />,
    color: '#8c8c8c',
    ext: ['md', 'mdx'],
    key: 'markdown',
  },
  {
    icon: <FilePdfFilled />,
    color: '#ff4d4f',
    ext: ['pdf'],
    key: 'pdf',
  },
  {
    icon: <FilePptFilled />,
    color: '#ff6e31',
    ext: ['ppt', 'pptx'],
    key: 'ppt',
  },
  {
    icon: <FileWordFilled />,
    color: '#1677ff',
    ext: ['doc', 'docx'],
    key: 'word',
  },
  {
    icon: <FileZipFilled />,
    color: '#fab714',
    ext: ['zip', 'rar', '7z', 'tar', 'gz'],
    key: 'zip',
  },
  {
    icon: <VideoIcon />,
    color: '#ff4d4f',
    ext: VIDEO_EXT,
    key: 'video',
  },
  {
    icon: <AudioIcon />,
    color: '#ff6e31',
    ext: AUDIO_EXT,
    key: 'audio',
  },
  {
    icon: <JavaOutlined />,
    color: '#1677ff',
    ext: ['java'],
    key: 'java',
  },
  {
    icon: <JavaScriptOutlined />,
    color: '#fab714',
    ext: ['js'],
    key: 'javascript',
  },
  {
    icon: <PythonOutlined />,
    color: '#fab714',
    ext: ['py'],
    key: 'python',
  },
];

const DEFAULT_ICON = {
  icon: <FileTextFilled />,
  color: '#8c8c8c',
  ext: ['default'],
  key: 'default',
};

const FileCard: React.FC<FileCardProps> = (props) => {
  const {
    prefixCls: customizePrefixCls,
    style,
    styles = {},
    className,
    rootClassName,
    classNames = {},
    name,
    byte,
    size,
    description,
    icon: customIcon,
    src,
    mask,
    type: customType,
    onClick,
    ...restProps
  } = props;

  const domProps = pickAttrs(restProps, {
    attr: true,
    aria: true,
    data: true,
  });

  const { direction, getPrefixCls } = useXProviderContext();
  const prefixCls = getPrefixCls('file-card', customizePrefixCls);
  const contextConfig = useXComponentConfig('fileCard');

  const [hashId, cssVarCls] = useStyle(prefixCls);

  const mergedCls = classnames(
    prefixCls,
    contextConfig.className,
    className,
    rootClassName,
    classNames.root,
    hashId,
    cssVarCls,
    {
      [`${prefixCls}-rtl`]: direction === 'rtl',
    },
  );

  const [namePrefix, nameSuffix] = useMemo(() => {
    const nameStr = name || '';
    const match = nameStr.match(/^(.*)\.[^.]+$/);
    return match ? [match[1], nameStr.slice(match[1].length)] : [nameStr, ''];
  }, [name]);

  const [icon, iconColor] = useMemo(() => {
    if (typeof customIcon === 'string') {
      const match = PRESET_FILE_ICONS.find((item) => item.key === customIcon);
      if (match) {
        return [match.icon, match.color];
      }
    }
    for (const item of PRESET_FILE_ICONS) {
      if (matchExt(nameSuffix, item.ext)) {
        return [item.icon, item.color];
      }
    }
    return [DEFAULT_ICON.icon, DEFAULT_ICON.color];
  }, [nameSuffix]);

  const fileType = useMemo(() => {
    if (customType) {
      return customType;
    }
    if (matchExt(nameSuffix, IMAGE_EXT)) {
      return 'image';
    }
    if (matchExt(nameSuffix, AUDIO_EXT)) {
      return 'audio';
    }
    if (matchExt(nameSuffix, VIDEO_EXT)) {
      return 'video';
    }

    return 'file';
  }, [nameSuffix, customType]);

  let ContentNode: React.ReactNode = null;
  if (fileType === 'image') {
    const preview = mask ? { mask } : undefined;
    ContentNode = (
      <Image
        rootClassName={classnames(`${prefixCls}-image`, classNames.file)}
        width={styles?.file?.width}
        height={styles?.file?.height}
        style={styles.file}
        alt={name}
        src={src}
        preview={preview}
        {...(restProps as ImageProps)}
      />
    );
  } else if (fileType === 'video') {
    ContentNode = (
      <video
        src={src}
        controls
        className={classnames(`${prefixCls}-video`, classNames.file)}
        {...(restProps as React.JSX.IntrinsicElements['video'])}
      />
    );
  } else if (fileType === 'audio') {
    ContentNode = (
      <audio
        src={src}
        controls
        className={classnames(`${prefixCls}-audio`, classNames.file)}
        {...(restProps as React.JSX.IntrinsicElements['audio'])}
      />
    );
  } else {
    ContentNode = (
      <File
        prefixCls={prefixCls}
        name={namePrefix}
        ext={nameSuffix}
        size={size}
        byte={byte}
        description={description}
        icon={customIcon && typeof customIcon !== 'string' ? customIcon : icon}
        iconColor={iconColor}
        onClick={onClick}
        mask={mask}
        classNames={classNames}
        styles={styles}
      />
    );
  }

  return (
    <div
      {...domProps}
      className={mergedCls}
      style={{
        ...contextConfig.style,
        ...style,
        ...styles.root,
      }}
    >
      {ContentNode}
    </div>
  );
};

if (process.env.NODE_ENV !== 'production') {
  FileCard.displayName = 'FileCard';
}

export default FileCard;
