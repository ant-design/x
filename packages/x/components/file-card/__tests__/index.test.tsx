import React from 'react';

import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { render } from '../../../tests/utils';
import FileCard from '../index';

describe('FileCard Component', () => {
  mountTest(() => <FileCard />);

  rtlTest(() => <FileCard />);

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('FileCard component work', () => {
    const { container } = render(<FileCard name="test-file.txt" size={1024} />);
    const element = container.querySelector<HTMLDivElement>('.ant-file-card');
    expect(element).toBeTruthy();
    expect(element).toMatchSnapshot();
  });

  it('FileCard support name and size', () => {
    const { container } = render(<FileCard name="test-file.txt" size={1024} />);
    const namePrefix = container.querySelector<HTMLDivElement>('.ant-file-card-file-name-prefix');
    expect(namePrefix?.textContent).toBe('test-file');
    const nameSuffix = container.querySelector<HTMLDivElement>('.ant-file-card-file-name-suffix');
    expect(nameSuffix?.textContent).toBe('.txt');
    const size = container.querySelector<HTMLDivElement>('.ant-file-card-file-description');
    expect(size?.textContent).toBe('1 KB');
  });

  it('FileCard support description', () => {
    const { container } = render(
      <FileCard name="test-file.txt" description={'test description'} />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-file-card-file-description');
    expect(element?.textContent).toBe('test description');
  });

  it('FileCard support image', () => {
    const { container } = render(
      <FileCard
        name="image-file.png"
        src="https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png"
      />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-file-card-image');
    expect(element).toBeTruthy();
  });

  it('FileCard support mask', () => {
    const { container } = render(
      <FileCard name="test-file.txt" size={1024} mask={<div className="test-mask" />} />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-file-card-file-mask .test-mask');
    expect(element).toBeTruthy();
  });

  it('FileCard support type', () => {
    const { container } = render(
      <FileCard
        name="image-file.png"
        src="https://zos.alipayobjects.com/rmsportal/jkjgkEfvpUPVyRjUImniVslZfWPnJuuZ.png"
        type="file"
      />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-file-card-image');
    expect(element).toBeFalsy();
    const file = container.querySelector<HTMLDivElement>('.ant-file-card-file');
    expect(file).toBeTruthy();
  });
});
