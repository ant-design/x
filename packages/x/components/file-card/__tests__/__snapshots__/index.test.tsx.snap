// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileCard Component FileCard component work 1`] = `
<div
  class="ant-file-card"
>
  <div
    class="ant-file-card-file"
  >
    <div
      class="ant-file-card-file-icon"
      style="color: #8c8c8c;"
    >
      <span
        aria-label="file-text"
        class="anticon anticon-file-text"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="file-text"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M854.6 288.7c6 6 9.4 14.1 9.4 22.6V928c0 17.7-14.3 32-32 32H192c-17.7 0-32-14.3-32-32V96c0-17.7 14.3-32 32-32h424.7c8.5 0 16.7 3.4 22.7 9.4l215.2 215.3zM790.2 326L602 137.8V326h188.2zM320 482a8 8 0 00-8 8v48a8 8 0 008 8h384a8 8 0 008-8v-48a8 8 0 00-8-8H320zm0 136a8 8 0 00-8 8v48a8 8 0 008 8h184a8 8 0 008-8v-48a8 8 0 00-8-8H320z"
          />
        </svg>
      </span>
    </div>
    <div
      class="ant-file-card-file-content"
    >
      <div
        class="ant-file-card-file-name"
      >
        <span
          class="ant-file-card-file-name-prefix"
        >
          test-file
        </span>
        <span
          class="ant-file-card-file-name-suffix"
        >
          .txt
        </span>
      </div>
      <div
        class="ant-file-card-file-description"
      >
        1 KB
      </div>
    </div>
  </div>
</div>
`;

exports[`FileCard Component rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-file-card ant-file-card-rtl"
>
  <div
    class="ant-file-card-file"
  >
    <div
      class="ant-file-card-file-icon"
      style="color: #8c8c8c;"
    >
      <span
        aria-label="file-text"
        class="anticon anticon-file-text"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="file-text"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M854.6 288.7c6 6 9.4 14.1 9.4 22.6V928c0 17.7-14.3 32-32 32H192c-17.7 0-32-14.3-32-32V96c0-17.7 14.3-32 32-32h424.7c8.5 0 16.7 3.4 22.7 9.4l215.2 215.3zM790.2 326L602 137.8V326h188.2zM320 482a8 8 0 00-8 8v48a8 8 0 008 8h384a8 8 0 008-8v-48a8 8 0 00-8-8H320zm0 136a8 8 0 00-8 8v48a8 8 0 008 8h184a8 8 0 008-8v-48a8 8 0 00-8-8H320z"
          />
        </svg>
      </span>
    </div>
    <div
      class="ant-file-card-file-content"
    >
      <div
        class="ant-file-card-file-name"
      >
        <span
          class="ant-file-card-file-name-prefix"
        />
        <span
          class="ant-file-card-file-name-suffix"
        />
      </div>
    </div>
  </div>
</div>
`;
